### DISCUSSION TITLE
make test fails on win10,built in source code from rdkit-Release_2021_09_4.zip

### DISCUSSION THREAD

**Discussion by hbsjzjzl** [rdkit/rdkit#5016] [Q&A] [answered]
I'm working in vs2022, python3.9, postgreSQL12.
When I run ctest,  I find almost all tests which contain "py" fail(total 204, and 66 fail). 
I find there's an unusual info in my cmake logs : "pthread.h not found".
I don't know whether if this cause my ctest fail 33%.
I hopefully want to know what's going wrong and what can I deal with it.
Who can help me, please!

My cmake command is as below:

cmake -DRDK_BUILD_PYTHON_WRAPPERS=ON -DBOOST_ROOT=D:/local/boost_1_78_0 -DBOOST_LIBRARYDIR=D:/local/boost_1_78_0/stage/lib -DBOOST_INCLUDEDIR=D:/local/boost_1_78_0/boost -DFREETYPE_LIBRARY=C:/freetype-2.11.1/objs/x64/Debug/freetype.lib -DFREETYPE_INCLUDE_DIRS=C:/freetype-2.11.1/include -DEIGEN3_INCLUDE_DIR="C:\Program Files\Eigen3\include\eigen3" -DZLIB_INCLUDE_DIR="D:\local\zlib\zlib-1.2.11_by_64" -DBoost_DEBUG=ON -DRDK_BUILD_INCHI_SUPPORT=ON -DRDK_BUILD_AVALON_SUPPORT=ON -DRDK_BUILD_PGSQL=ON -DPostgreSQL_ROOT="D:\Program Files\PostgreSQL\12" -G"Visual Studio 17 2022" ..

I pick up one name's "pythonTestDirChem" and get the info as below:

Test: pythonTestDirChem
Command: "D:/Program Files (x86)/Microsoft Visual Studio/Shared/Python39_64/python.exe" "D:/RDKit/rdkit/Chem/test_list.py" "--testDir" "D:/RDKit/rdkit/Chem" "--buildType" "Release"
Directory: D:/RDKit/build/rdkit/Chem
"pythonTestDirChem" start time: Feb 17 17:05 ?D1ú±ê×?ê±??
Output:
----------------------------------------------------------
Traceback (most recent call last):
  File "D:\RDKit\rdkit\Chem\test_list.py", line 60, in <module>
    from rdkit import TestRunner
  File "D:\RDKit\rdkit\__init__.py", line 2, in <module>
    from . import rdBase
ImportError: DLL load failed while importing rdBase: 找不到指定的模块。
<end of output>
Test time =   0.05 sec
----------------------------------------------------------
Test Failed.
"pythonTestDirChem" end time: Feb 17 17:05 ?D1ú±ê×?ê±??
"pythonTestDirChem" time elapsed: 00:00:00

**Comment by ptosco** [comment, marked as answer]
1. You need to make sure that the boost DLLs in `D:\local\boost_1_78_0\stage\lib` are in your `PATH`
2. You are running tests with `D:/Program Files (x86)/Microsoft Visual Studio/Shared/Python39_64/python.exe`. Is this the same Python interpreter that you have built boost and RDKit against? You can check this in the boost build logs and in the RDKit `cmake` log. I notice that you have not specified `PYTHON_EXECUTABLE`, `PYTHON_INCLUDE_DIR` and `PYTHON_LIBRARY` in your RDKit `cmake` command, so `cmake` may have found a Python interpreter on your system which is not `D:/Program Files (x86)/Microsoft Visual Studio/Shared/Python39_64/python.exe`. The same might have happened for boost. This would explain why all Python tests are failing. You need to make sure that you use the same Python interpreter to build boost, RDKit and to run `ctest`.

**Reply by hbsjzjzl** [reply]
Thanks a lot, it's really helpful! After checking my python version and python setting from boost, cmake, msbuild to ctest, only three tests fail and two tests not run now(finally i use python 3.7.3 instead). I wonder if you know why these three tests still fail.
During my msbuild progress,there are many warning contains "Unicode" or what ,sorry i do not get the msbuild log at the first time, and at the second time,warning disappeared, even i don't change anything.
my new cmake command is as below:

cmake -DRDK_BUILD_PYTHON_WRAPPERS=ON -DPYTHON_EXECUTABLE="D:\Programs\Python\Python37\python.exe" -DPYTHON_INCLUDE_DIR="D:\Programs\Python\Python37\include" -DBOOST_ROOT=D:/local/boost_1_78_0 -DBOOST_LIBRARYDIR=D:/local/boost_1_78_0/stage/lib -DBOOST_INCLUDEDIR=D:/local/boost_1_78_0/boost -DFREETYPE_LIBRARY=C:/freetype-2.11.1/objs/x64/Debug/freetype.lib -DFREETYPE_INCLUDE_DIRS=C:/freetype-2.11.1/include -DEIGEN3_INCLUDE_DIR="C:\Program Files\Eigen3\include\eigen3" -DZLIB_INCLUDE_DIR="D:\local\zlib\zlib-install\include" -DZLIB_LIBRARY="D:\local\zlib\zlib-install\lib\zlibstaticd.lib" -DBoost_DEBUG=ON -DRDK_BUILD_INCHI_SUPPORT=ON -DRDK_BUILD_AVALON_SUPPORT=ON -DRDK_BUILD_PGSQL=ON -DPostgreSQL_ROOT="D:\Program Files\PostgreSQL\12" -G"Visual Studio 17 2022" ..

and three failed ctest info is as below:

178 - pyMolStandardize (Failed)
File "D:/RDKit/Code/GraphMol/MolStandardize/Wrap/testMolStandardize.py", line 15, in <module>
    from rdkit.Chem.MolStandardize import rdMolStandardize
  File "D:\RDKit\rdkit\Chem\MolStandardize\__init__.py", line 17, in <module>
    from rdkit.Chem.MolStandardize import rdMolStandardize
UnicodeDecodeError: 'utf-8' codec can't decode bytes in position 167-168: invalid continuation byte

195 - testPgSQL (Failed)
D:\RDKit\build\Code\PgSQL\rdkit>"D:\Program Files\PostgreSQL\12\bin\pg_regress.exe" --inputdir=sql --bindir="D:/Program Files/PostgreSQL/12/bin" rdkit-91 props btree molgist bfpgist-91 sfpgist slfpgist fps reaction fmcs avalon inchi json 
(using postmaster on localhost, default port)
============== dropping database "regression"         ==============
DROP DATABASE
============== creating database "regression"         ==============
CREATE DATABASE
ALTER DATABASE
============== running regression test queries        ==============
test rdkit-91                     ... 'diff' 不是内部或外部命令，也不是可运行的程序
或批处理文件。
diff command not found: diff -w "D:/RDKit/build/Code/PgSQL/rdkit/expected/rdkit-91.out" "D:/RDKit/build/Code/PgSQL/rdkit/results/rdkit-91.out" > "D:/RDKit/build/Code/PgSQL/rdkit/results/rdkit-91.out.diff"
<end of output>
Test time =   4.83 sec
----------------------------------------------------------
Test Failed.

203 - pythonTestDirChem (Failed)
203
File "UnitTestMolStandardize.py", line 10, in <module>
    from rdkit.Chem import MolStandardize
  File "D:\RDKit\rdkit\Chem\MolStandardize\__init__.py", line 17, in <module>
    from rdkit.Chem.MolStandardize import rdMolStandardize
UnicodeDecodeError: 'utf-8' codec can't decode bytes in position 167-168: invalid continuation byte

I don't understand ctest 195 trying to use diff command which is not contained in win10.
And I do want to use rdkit.Chem.MolStandardize,  but I can't figure out what's wrong.

**Comment by hbsjzjzl** [comment]
您好，您的邮件已收到。

**Comment by ptosco** [comment]
The `UnicodeDecodeError` raised by tests 178 and 203 could be due to the code page/locale.
You might try changing the code page in the shell where you are running `ctest` (e.g., `chcp 437`) and see if that fixes the problem.
For 195, you may find `diff` Windows implementations in MSYS2 or GnuWin32.
It is also worth mentioning that an up-to-date, pre-built RDKit Windows distribution is available through `conda-forge`. You may find extensive information to install it on the web.

**Reply by hbsjzjzl** [reply]
Conda is really  convinient for me to start rdkit. Everything goes very in conda. Although I still do not fix the "UnicodeDecodeError" prolem and  souce ctest codes are in rdkit source code. Maybe some tools or some versions I installed manually are not matched. Anyway, thanks a lot~
